### ISSUE TITLE
Potential memory leak during Llama 3 8b model fine-tuning with LoRA

### ISSUE THREAD

**Issue by mzbac** [ml-explore/mlx-examples#738]
While trying to fine-tune `meta-llama/Meta-Llama-3-8B-Instruct` during training, the peak memory slowly grew to around 300GB and eventually caused the mlx-lm crashes.

**How to reproduce**
- mlx 0.11.0
- training dataset from https://huggingface.co/datasets/mzbac/function-calling-llama-3-format-v1.1/tree/main
- lora_config.yaml
```yaml
# The path to the local model directory or Hugging Face repo.
model: "meta-llama/Meta-Llama-3-8B-Instruct"
# Whether or not to train (boolean)
train: true

# Directory with {train, valid, test}.jsonl files
data: "data"

# The PRNG seed
seed: 0

# Number of layers to fine-tune
lora_layers: 32

# Minibatch size.
batch_size: 1

# Iterations to train for.
iters: 6000

# Number of validation batches, -1 uses the entire validation set.
val_batches: 25

# Adam learning rate.
learning_rate: 1e-6

# Number of training steps between loss reporting.
steps_per_report: 10

# Number of training steps between validations.
steps_per_eval: 200

# Load path to resume training with the given adapter weights.
resume_adapter_file: null

# Save/load path for the trained adapter weights.
adapter_path: "adapters"

# Save the model every N iterations.
save_every: 1000

# Evaluate on the test set after training
test: false

# Number of test set batches, -1 uses the entire test set.
test_batches: 100

# Maximum sequence length.
max_seq_length: 8192

# Use gradient checkpointing to reduce memory use.
grad_checkpoint: true

# LoRA parameters can only be specified in a config file
lora_parameters:
  # The layer keys to apply LoRA to.
  # These will be applied for the last lora_layers
  keys: ['mlp.gate_proj', 'mlp.down_proj', 'self_attn.q_proj', 'mlp.up_proj', 'self_attn.o_proj','self_attn.v_proj', 'self_attn.k_proj']
  rank: 128
  alpha: 256
  scale: 10.0
  dropout: 0.05

# Schedule can only be specified in a config file, uncomment to use.
# lr_schedule:
#  name: cosine_decay
#  warmup: 100 # 0 for no warmup
#  warmup_init: 1e-7 # 0 if not specified
#  arguments: [1e-6, 1000, 1e-7] # passed to scheduler
```

**logs**
```
Iter 6010: Train loss 0.430, Learning Rate 1.000e-06, It/sec 0.106, Tokens/sec 47.476, Trained Tokens 3213544, Peak mem 289.013 GB
Iter 6020: Train loss 0.313, Learning Rate 1.000e-06, It/sec 0.109, Tokens/sec 49.663, Trained Tokens 3218096, Peak mem 289.013 GB
Iter 6030: Train loss 0.503, Learning Rate 1.000e-06, It/sec 0.085, Tokens/sec 61.099, Trained Tokens 3225260, Peak mem 289.013 GB
Iter 6040: Train loss 0.332, Learning Rate 1.000e-06, It/sec 0.108, Tokens/sec 49.178, Trained Tokens 3229793, Peak mem 289.013 GB
Iter 6050: Train loss 0.268, Learning Rate 1.000e-06, It/sec 0.104, Tokens/sec 52.523, Trained Tokens 3234820, Peak mem 289.013 GB
```
<img width="1272" alt="image" src="https://github.com/ml-explore/mlx-examples/assets/7523197/ff55e198-2684-44fd-90b9-7a2d741c21e4">

**Comment by awni** [comment]
Very strange .. seems like there is a leak of some sort.

**Comment by mzbac** [comment]
and the crash logs:
```
Iter 7320: Train loss 0.216, Learning Rate 1.000e-06, It/sec 0.116, Tokens/sec 40.214, Trained Tokens 3893259, Peak mem 342.467 GB
libc++abi: terminating due to uncaught exception of type std::runtime_error: [malloc_or_wait] Unable to allocate 554319936 bytes.
zsh: abort      mlx_lm.lora --config lora_config.yaml
```

**Comment by awni** [comment]
Seems to be related to gradient checkpointing but that's all I know so far.. turning off gradient checkpointing and the peak memory is constant for a fixed sequence length as expected.

**Comment by mzbac** [comment]
> Seems to be related to gradient checkpointing but that's all I know so far.. turning off gradient checkpointing and the peak memory is constant for a fixed sequence length as expected.

After disabling gradient checkpointing, the peak memory has stabilized. I will continue training for a few thousand iterations to see if everything is running as expected. Thanks @awni 🚀

**Comment by Satyam7166-tech** [comment]
@mzbac, a bit off topic and please let me know if I'm breaking some rules so I'll delete the comment.

But can you share an example of your dataset used for finetuning llama3 instruct? I am a bit confued by its template.

Thanks

**Comment by mzbac** [comment]
> @mzbac, a bit off topic and please let me know if I'm breaking some rules so I'll delete the comment.
> 
> But can you share an example of your dataset used for finetuning llama3 instruct? I am a bit confued by its template.
> 
> Thanks

@Satyam7166-tech  You can check the data here at https://huggingface.co/datasets/mzbac/function-calling-llama-3-format-v1.1/viewer and feel free to create a discussion there if you have any questions about the dataset.

**Comment by mzbac** [comment]
@awni I can confirm that the issue is related to gradient checkpointing. With gradient checkpointing disabled, I trained the model for 8000 iterations and the peak memory is consistent and works as expected.

**Comment by awni** [comment]
Just FYI, the leak should be fixed now with gradient checkpointing enabled.

**Comment by ivanfioravanti** [comment]
Thanks for the fix!
